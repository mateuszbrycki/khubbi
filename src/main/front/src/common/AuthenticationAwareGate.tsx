import React from "react";
import {Navigate} from "react-router-dom";


export interface AuthRouteProps {
    readonly children: React.ReactNode;
    readonly isAuthenticated: () => boolean;
    readonly redirectToIfLogged?: string;
    readonly redirectToIfNotLogged?: string;
}


export const AuthenticationAwareGate: React.FC<AuthRouteProps> = ({
                                                                      isAuthenticated,
                                                                      children,
                                                                      redirectToIfLogged,
                                                                      redirectToIfNotLogged
                                                                  }) => {

    if (!isAuthenticated()) {
        if (redirectToIfNotLogged) {
            return <Navigate to={redirectToIfNotLogged}/>;
        }
        // TODO mateusz.brycki this causes problem for the login page - the infinite redirection loop
        // return <Navigate to="/login"/>;
    }

    if (isAuthenticated() && redirectToIfLogged) {
        return <Navigate to={redirectToIfLogged}/>;
    }

    return <>{children}</>;
};


export default AuthenticationAwareGate