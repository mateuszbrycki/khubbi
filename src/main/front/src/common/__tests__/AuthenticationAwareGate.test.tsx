import ShallowRenderer from "react-test-renderer/shallow";
import AuthenticationAwareGate from "../AuthenticationAwareGate";

describe('authentication aware gate', () => {
    describe('when user is authenticated', () => {
        const isAuthenticated = true

        it('renders test child without redirections', () => {
            const renderer = ShallowRenderer.createRenderer()
            renderer.render(<AuthenticationAwareGate children={<h1>test child</h1>}
                                                     isAuthenticated={() => isAuthenticated}/>)
            expect(renderer.getRenderOutput()).toMatchSnapshot()
        })

        it('renders redirects to dashboard', () => {
            const renderer = ShallowRenderer.createRenderer()
            renderer.render(<AuthenticationAwareGate children={<h1>test child</h1>}
                                                     isAuthenticated={() => isAuthenticated}
                                                     redirectToIfLogged={"/dashboard"}/>)
            expect(renderer.getRenderOutput()).toMatchSnapshot()
        })
    });

    describe('when user is not authenticated', () => {
        const isAuthenticated = false

        it('navigates to login', () => {
            const renderer = ShallowRenderer.createRenderer()
            renderer.render(<AuthenticationAwareGate children={<h1>test child</h1>}
                                                     isAuthenticated={() => isAuthenticated}/>)
            expect(renderer.getRenderOutput()).toMatchSnapshot()
        })

        it('navigates to given url', () => {
            const renderer = ShallowRenderer.createRenderer()
            renderer.render(<AuthenticationAwareGate children={<h1>test child</h1>}
                                                     isAuthenticated={() => isAuthenticated}
                                                     redirectToIfNotLogged={"/login-test"}/>)
            expect(renderer.getRenderOutput()).toMatchSnapshot()
        })
    });
})
