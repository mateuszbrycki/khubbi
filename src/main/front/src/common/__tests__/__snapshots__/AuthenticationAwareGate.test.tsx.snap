// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authentication aware gate when user is authenticated renders redirects to dashboard 1`] = `
<Navigate
  to="/dashboard"
/>
`;

exports[`authentication aware gate when user is authenticated renders test child without redirections 1`] = `
<React.Fragment>
  <h1>
    test child
  </h1>
</React.Fragment>
`;

exports[`authentication aware gate when user is not authenticated navigates to given url 1`] = `
<Navigate
  to="/login-test"
/>
`;

exports[`authentication aware gate when user is not authenticated navigates to login 1`] = `
<React.Fragment>
  <h1>
    test child
  </h1>
</React.Fragment>
`;
